services/api.py

```diff
@@ -816,20 +816,38 @@ def service_nodes_by_ancestors(service_node_ids):
                 bbox_geometry_filter = munigeo_api.build_bbox_filter(ref, val, 'geometry')
                 queryset = queryset.filter(Q(**bbox_filter) | Q(**bbox_geometry_filter))
 
+        def get_category_ids(categories, category):
+            ids = []
+            for mapping in categories:
+                if mapping.split(':')[0] == category:
+                    ids.append(mapping.split(':')[1])
+            return ids
+
         if 'category' in filters:
             services_and_service_nodes = filters.get('category', None).split(',')
-            service_ids = []
-            servicenode_ids = []
-            for category in services_and_service_nodes:
-                key = category.split(':')[0]
-                value = category.split(':')[1]
-                if key == 'service':
-                    service_ids.append(value)
-                elif key == 'service_node':
-                    servicenode_ids.append(value)
+            service_ids = get_category_ids(services_and_service_nodes, 'service')
+            servicenode_ids = get_category_ids(services_and_service_nodes, 'service_node')
             queryset = queryset.filter(Q(services__in=service_ids)
                                        | Q(service_nodes__in=service_nodes_by_ancestors(servicenode_ids))).distinct()
 
+        if 'period' in filters:
+            period = filters.get('period', None)
+            services = None
+            if 'service' in filters:
+                services = filters.get('service').split(',')
+            if 'category' in filters:
+                categories = filters.get('category', None).split(',')
+                services = get_category_ids(categories, 'service')
+
+            if services is not None:
+                unit_service_details = UnitServiceDetails.objects.filter(service__in=services).distinct()\
+                    .filter(Q(period_begin_year=period) | Q(period_begin_year=None))
+                units = list(map(lambda x: x.unit.id, unit_service_details))
```

Comment:
This code is okay, and correct, but using list comprehensions would be more idiomatic Python:

    [x.unit.id for x in unit_service_details]

However, see the next comment before implementing this.

https://www.pythonforbeginners.com/basics/list-comprehensions-in-python